### DISCUSSION TITLE
How to provide configuration to the Op inside of a Graph Asset?

### DISCUSSION THREAD

**Discussion by marcilj** [dagster-io/dagster#19570] [Q&A]
Hey everyone, I want to provide all of my Op in a Graph Asset the same configuration and I just can’t figure it out.
Here’s what I’ve tried.
So I’ve started with a simple graph asset that have 2 Op in it.
## Step 1 : Graph Asset without config

```python
@op
def op_1():
  return 1


@op
def op_2(number: int):
  return number * 2


@graph_asset
def graph_asset_1():
  return op_2(op_1())
```
Then I’ve added a job that would have my graph asset in it.
## Step 2 : Graph Asset with job (That doesn’t work)

```python
@op
def op_1():
  return 1


@op
def op_2(number: int):
  return number * 2


@graph_asset()
def graph_asset_1():
  return op_2(op_1())


@job()
def job_1():
  graph_asset_1()
```
After some research it seems like we can’t call the Asset directly in a job define with the job decorator. We need to use the function define_asset_job.
Here’s the code with a job that work.
## Step 3 : Graph Asset with job that work
```python
@op
def op_1():
  return 1


@op
def op_2(number: int):
  return number * 2


@graph_asset()
def graph_asset_1():
  return op_2(op_1())


job_1 = define_asset_job("job_1", selection=["graph_asset_1"])
```
## Step 4 : Graph Asset with config from job
```python
class MyOpConfig(ConfigurableResource):
  a_int: int


@op
def op_1(config: MyOpConfig):
  return config.a_int


@op
def op_2(number: int):
  return number * 2


@graph_asset()
def graph_asset_1():
  return op_2(op_1())


job_1 = define_asset_job(
  "job_1",
  selection=["graph_asset_1"],
  config=RunConfig(ops={"op_1": MyOpConfig(a_int=2)}),
)
```
Doesn’t work. Receive the following error.
dagster._core.errors.DagsterInvalidDefinitionError: resource with key 'config' required by op 'op_1' was not provided.

In the end I would like my 2 ops to have the same `config: MyOpConfig` configured, but I've started with configuring 1 at the time to see where things doesn't work, and It ended there.

## Step 5 : Graph Asset with config from job
I’ve found that thread https://dagster.slack.com/archives/C01U954MEER/p1699986775907899?thread_ts=1699982827.893069&cid=C01U954MEER
And in that thread you configure the config in the `RunConfig(resources)` instead of in the `RunConfig(ops)` so I’ve tried it.
```python
class MyOpConfig(ConfigurableResource):
  a_int: int


@op
def op_1(config: MyOpConfig):
  return config.a_int


@op
def op_2(number: int):
  return number * 2


@graph_asset()
def graph_asset_1():
  return op_2(op_1())


job_1 = define_asset_job(
  "job_1",
  selection=["graph_asset_1"],
  config=RunConfig(resources=MyOpConfig(a_int=2)),
)
```
Still doesn’t work. I received that exception:
The above exception was caused by the following exception:
```
dagster._check.ParameterCheckError: Param "resources" is not a dict. Got MyOpConfig(a_int=2, _state__internal__=ConfigurableResourceFactoryState(nested_partial_resources={}, resolved_config_dict={'a_int': 2}, config_schema=<dagster._core.definitions.definition_config_schema.DefinitionConfigSchema object at 0x128031b20>, schema=Field(<dagster._config.field_utils.Shape object at 0x128031880>, default=@, is_required=True), nested_resources={}, resource_context=None)) which is type <class 'my_dagster_project.ops.MyOpConfig'>.
```
Step 6 :
Then I found this thread https://github.com/dagster-io/dagster/discussions/3213#discussioncomment-7569365
Adding a `Dict` in `RunConfig` didn’t help either.
```python
class MyOpConfig(ConfigurableResource):
  a_int: int


@op
def op_1(config: MyOpConfig):
  return config.a_int


@op
def op_2(number: int):
  return number * 2


@graph_asset()
def graph_asset_1():
  return op_2(op_1())


job_1 = define_asset_job(
  "job_1",
  selection=["graph_asset_1"],
  config=RunConfig(resources={"config": MyOpConfig(a_int=2)}),
)
```
Error Received:

```
dagster._core.errors.DagsterInvalidDefinitionError: resource with key 'config' required by op 'op_1' was not provided.
Also replicating the code in that thread (Changing the graph asset to a graph) did make Dagster start, but the config was bad.
```
```python
class MyOpConfig(ConfigurableResource):
  a_int: int

@op
def op_1(config: MyOpConfig):
  return config.a_int


@op
def op_2(number: int):
  return number * 2


@graph()
def graph_asset_1():
  return op_2(op_1())


job_1 = graph_asset_1.to_job(resource_defs={"config": MyOpConfig(a_int=2)})
```
Configuration in the UI
```yml
resources:
 config:
  config:
   a_int: 2
```
Vs Scaffold Missing Config
```yml
ops:
 op_1:
  config:
   a_int: 0
```
So here I am. Still not able to provide any configuration to my Op in a Graph Asset.
Anyway you can point me to where I did something wrong?
Also am I the only one who’s using Graph Assets? :sweat_smile:

**Comment by alangenfeld** [comment]
A couple things 
* the location of the ops config gets nested when its contained within a `graph` https://docs.dagster.io/concepts/ops-jobs-graphs/nesting-graphs#sub-graph-configuration
* you need to use the `Config` class for config , your examples were using `ConfigurableResource` https://docs.dagster.io/concepts/configuration/config-schema

Here is one example I tested as successful: 
```python
from dagster import define_asset_job, op, graph_asset, Config, Definitions, RunConfig


class MyOpConfig(Config):
    a_int: int


@op()
def op_1(config: MyOpConfig):
    return config.a_int


@op
def op_2(number: int):
    return number * 2


@graph_asset()
def graph_asset_1():
    return op_2(op_1())


job_1 = define_asset_job(
    "job_1",
    selection=["graph_asset_1"],
    config=RunConfig(
        ops={
            "graph_asset_1": {
                "ops": {
                    "op_1": MyOpConfig(a_int=2),
                }
            }
        }
    ),
)
defs = Definitions(
    jobs=[job_1],
    assets=[graph_asset_1],
)
```

For your root problem, you may want to use the `config` argument on the `graph_asset` directly which can accept a fixed config or a `@config_mapping` function https://docs.dagster.io/_apidocs/assets#dagster.graph_asset

```python
from dagster import define_asset_job, op, graph_asset, Config, Definitions


class MyOpConfig(Config):
    a_int: int


@op()
def op_1(config: MyOpConfig):
    return config.a_int


@op
def op_2(number: int):
    return number * 2


@graph_asset(
    config={
        "op_1": {
            # current bug: this doesn't handle MyOpConfig so we have to pass raw config
            "config": {"a_int": 2}
        },
    }
)
def graph_asset_1():
    return op_2(op_1())


job_1 = define_asset_job(
    "job_1",
    selection=["graph_asset_1"],
)
defs = Definitions(
    jobs=[job_1],
    assets=[graph_asset_1],
)
```

**Reply by marcilj** [reply]
Thank you very much @alangenfeld for the example. This helps me a lot.

In the provided example, one thing that bug me is the fact that I need to provide a config for`all` the Op.

In this example if I have 3 Op in my Graph Asset that need the config, I would need to configure my GraphAsset with a configuration that look like this

```python
@graph_asset(
    config={
        "op_1": {
            # current bug: this doesn't handle MyOpConfig so we have to pass raw config
            "config": {"a_int": 2}
        },
        "op_2": {
            # current bug: this doesn't handle MyOpConfig so we have to pass raw config
            "config": {"a_int": 2}
        },
        "op_3": {
            # current bug: this doesn't handle MyOpConfig so we have to pass raw config
            "config": {"a_int": 2}
        },
    }
)
```

What I want do to is configure a set of configuration at the job level that all my Op would be able to access.

After browsing the ask-community channel I ended up on a [thread](https://dagster.slack.com/archives/C01U954MEER/p1699986944457729?thread_ts=1699982827.893069&cid=C01U954MEER) were somewhere from Dagster was suggesting using `ConfigurableResource` to pass the `same` configuration to multiple Ops since `ConfigurableResource` inherits from `Config`.

Note: I have around 6 similars queries that I need to do for 10 differents Stocks. I've started creating classes that inherits from others and ended up with a working code. The issue is that all of my code was wrapper in python classes and I was only calling a bunch of function in a Dagster Asset. That created some problem since I didn't had a good visibility on what was happening, retries were handled by me in the code, and not by dagster, and each steps were like a big one from Dagster point of view. What I'm trying to do now is convert that to leverage Dagster built in feature, so I am splitting all the classes and elements of those classes into Ops. I want to wrap all of those in a Graph Asset since I want to materialize a table at the end step. And finally, since most of the elements do similar things, I want to ease the addition of new stocks as much as possible. The way I think of doing this is by providing a set of variables that all my Ops could use in order to do what they should. I think of this like a class and It might be wrong, but my understanding is that I could create a resource that hold a config with a bunch of parameters that are linked to a stock. Then pass that around in all my Ops. Then adding a new Stock to my list would only require the creation of a new job with a different configuration.

**Reply by alangenfeld** [reply]
Here is a reference to a discussion around using a resource to share config among many nodes https://github.com/dagster-io/dagster/discussions/3213

I think you are in the thick of working through some real design decisions that require full context to navigate. My vague advice would be to consider what exactly you are configuring and how you could factor that information. For example, if some common settings are related to how to perform the "6 similar queries" maybe you group those settings in to one resource or push the construction/execution of those queries down in to a shared resource.
